system/service/auth_credentials.go

```diff
@@ -0,0 +1,1157 @@
+	var (
+		expiresAt time.Time
+		aam       = &authActionProps{
+			user:        u,
+			credentials: &types.Credentials{Kind: kind},
+		}
+	)
+
+	err = svc.store.Tx(ctx, func(ctx context.Context, s store.Storer) (err error) {
+		if u == nil || u.ID == 0 {
+			return AuthErrGeneric()
+		}
+
+		// Rate limit requests
+		cc, _, err := store.SearchCredentials(ctx, s, types.CredentialsFilter{
+			OwnerID: u.ID,
+			Kind:    kind,
+
+			// we want to count deleted tokens as well
+			Deleted: filter.StateInclusive,
+		})
+
+		if err != nil {
+			return err
+		}
+
+		// gt/eq since this current request is not yet stored
+		if err = svc.checkTokenRate(cc, tokenReqMaxWindow, tokenReqMaxCount); err != nil {
+			return
+		}
+
+		// removes expired and soft-deleted tokens
+		// and enforces one-token-per-user rule
+		if err = svc.cleanupCredentials(ctx, s, cc); err != nil {
+			return
+		}
+
+		switch kind {
+		case credentialsTypeMFAEmailOTP:
+			expSec := svc.settings.Auth.MultiFactor.EmailOTP.Expires
+			if expSec == 0 {
+				expSec = 60
+			}
+
+			expiresAt = now().Add(time.Second * time.Duration(expSec))
+
+			// random number, 6 chars
+			token = fmt.Sprintf("%06d", rand2.Int())[0:6]
+		case credentialsTypeCreatePasswordToken:
+			expSec := svc.settings.Auth.Internal.PasswordCreate.Expires
+			if expSec == 0 {
+				expSec = 24
+			}
+
+			expiresAt = now().Add(time.Hour * time.Duration(expSec))
+
+			// random password string, "3i[g0|)z"
+			token = fmt.Sprintf("%s", rand.Password(credentialsTokenLength))
+		default:
+			// 1h expiration for all tokens send via email
+			expiresAt = now().Add(time.Minute * 60)
+			token = string(rand.Bytes(credentialsTokenLength))
+		}
+
+		c := &types.Credentials{
+			ID:          nextID(),
+			CreatedAt:   *now(),
+			OwnerID:     u.ID,
+			Kind:        kind,
+			Credentials: token,
+			ExpiresAt:   &expiresAt,
+		}
+
+		err = store.CreateCredentials(ctx, s, c)
+
+		if err != nil {
+			return err
+		}
+
+		switch kind {
+		case credentialsTypeMFAEmailOTP:
+			// do not alter the final token
+		default:
+			// suffixing tokens with credentials ID
+			// this will help us with token lookups
+			token = fmt.Sprintf("%s%d", token, c.ID)
+		}
+
+		return nil
+	})
+
+	return token, svc.recordAction(ctx, aam, AuthActionIssueToken, err)
+}
+
+// checks existing tokens and ensure that the creation rate is within limits
+func (svc *auth) checkTokenRate(cc types.CredentialsSet, window time.Duration, max int) error {
+	if len(cc) == 0 || window == 0 || max == 0 {
+		return nil
+	}
+
+	var (
+		cutoff = now().Add(window * -1)
+		count  = 0
+	)
+
+	for _, c := range cc {
+		if c.CreatedAt.Before(cutoff) {
+			// skip tokens created before cutoff
+			continue
+		}
+
+		count++
+
+		if count > max {
+			break
+		}
+	}
+
+	if count > max {
+		return AuthErrRateLimitExceeded()
+	}
+
+	return nil
+}
+
+func (svc *auth) cleanupCredentials(ctx context.Context, s store.Credentials, cc types.CredentialsSet) (err error) {
```

Comment:
Won't this delete old user passwords making the password reuse check not possible?